Compiler accepts return mut ref to local var on no longer valid stackframe

Spotted on reddit: https://www.reddit.com/r/rust/comments/7i64sy/safe_function_returns_static_mut/  
credit to [jDomantas](https://www.reddit.com/user/jDomantas)

I tried this code: [playground](https://play.rust-lang.org/?gist=5e19e59e8b64488171b76bfbf0383b06&version=stable)

	fn gimme_static_mut() -> &'static mut u32 {
	    let ref mut x = 1234543;
	    x
	}

	fn main() {
	    let x = gimme_static_mut();
	    let y = gimme_static_mut();
	    *y = 42;
	    let a = *x;
	    println!("a: {}", a);
	}

This is accepted by the compiler and prints 42.

I expected to see this happen: The program should be rejected by the compiler with a lifetime error

Instead this happened: The program was accepted and miscompiled badly

The assembly code clearly demonstrates the problem, a local variable is created for the literal, its address is taken and returned:

	playground::gimme_static_mut:
	.Lfunc_begin2:
		push	rbp
	.Lcfi6:
	.Lcfi7:
		mov	rbp, rsp
	.Lcfi8:
		sub	rsp, 16
		lea	rax, [rbp - 4]
	.Ltmp6:
		mov	dword ptr [rbp - 4], 1234543
		mov	qword ptr [rbp - 16], rax
		mov	rax, qword ptr [rbp - 16]
	.Ltmp7:
		add	rsp, 16
		pop	rbp
		ret
	.Ltmp8:
	.Lfunc_end2:

## Meta

Latest version on playground

An alternative way to define gimme_static_mut:

	fn gimme_static_mut() -> &'static mut u32 {
	    let x = &mut 1234543;
	    x
	}

Gives the expected compiler error:

	   Compiling playground v0.0.1 (file:///playground)
	error[E0597]: borrowed value does not live long enough
	 --> src/main.rs:2:18
	  |
	2 |     let x = &mut 1234543;
	  |                  ^^^^^^^ does not live long enough
	3 |     x
	4 | }
	  | - temporary value only lives until here
	  |
	  = note: borrowed value must be valid for the static lifetime...

	error: aborting due to previous error

	error: Could not compile `playground`.

	To learn more, run the command again with --verbose.

Looks like an oversight with the `ref mut` pattern match syntax to capture the literal by reference.

--- a/src/test/ui/nll/borrowed-referent-issue-38899.rs
+++ b/src/test/ui/nll/borrowed-referent-issue-38899.rs
@@ -1,0 +1,32 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Regression test for issue #38899
+
+// compile-flags:-Znll -Zborrowck=mir -Zverbose
+
+#![feature(nll)]
+#![allow(dead_code)]
+
+pub struct Block<'a> {
+    current: &'a u8,
+    unrelated: &'a u8,
+}
+
+fn bump<'a>(mut block: &mut Block<'a>) {
+    let x = &mut block;
+    println!("{}", x.current);
+    let p: &'a u8 = &*block.current;
+    //~^ ERROR cannot borrow `*block.current` as immutable because it is also borrowed as mutable
+    drop(x);
+    drop(p);
+}
+
+fn main() {}

--- a/src/test/ui/nll/borrowed-referent-issue-38899.rs
+++ b/src/test/ui/nll/borrowed-referent-issue-38899.rs
@@ -10,8 +10,6 @@
 
 // Regression test for issue #38899
 
-// compile-flags:-Znll -Zborrowck=mir -Zverbose
-
 #![feature(nll)]
 #![allow(dead_code)]
 

--- a/src/test/ui/nll/borrowed-referent-issue-38899.stderr
+++ b/src/test/ui/nll/borrowed-referent-issue-38899.stderr
@@ -1,0 +1,11 @@
+error[E0502]: cannot borrow `*block.current` as immutable because it is also borrowed as mutable
+  --> $DIR/borrowed-referent-issue-38899.rs:26:21
+   |
+24 |     let x = &mut block;
+   |             ---------- mutable borrow occurs here
+25 |     println!("{}", x.current);
+26 |     let p: &'a u8 = &*block.current;
+   |                     ^^^^^^^^^^^^^^^ immutable borrow occurs here
+
+error: aborting due to previous error
+

--- a/src/test/ui/nll/borrowed-referent-issue-38899.stderr
+++ b/src/test/ui/nll/borrowed-referent-issue-38899.stderr
@@ -1,10 +1,10 @@
 error[E0502]: cannot borrow `*block.current` as immutable because it is also borrowed as mutable
-  --> $DIR/borrowed-referent-issue-38899.rs:26:21
+  --> $DIR/borrowed-referent-issue-38899.rs:24:21
    |
-24 |     let x = &mut block;
+22 |     let x = &mut block;
    |             ---------- mutable borrow occurs here
-25 |     println!("{}", x.current);
-26 |     let p: &'a u8 = &*block.current;
+23 |     println!("{}", x.current);
+24 |     let p: &'a u8 = &*block.current;
    |                     ^^^^^^^^^^^^^^^ immutable borrow occurs here
 
 error: aborting due to previous error

--- a/src/test/ui/nll/return-ref-mut-issue-46557.rs
+++ b/src/test/ui/nll/return-ref-mut-issue-46557.rs
@@ -10,8 +10,6 @@
 
 // Regression test for issue #46557
 
-// compile-flags:-Znll -Zborrowck=mir -Zverbose
-
 #![feature(nll)]
 #![allow(dead_code)]
 

--- a/src/test/ui/nll/return-ref-mut-issue-46557.rs
+++ b/src/test/ui/nll/return-ref-mut-issue-46557.rs
@@ -1,0 +1,23 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Regression test for issue #46557
+
+// compile-flags:-Znll -Zborrowck=mir -Zverbose
+
+#![feature(nll)]
+#![allow(dead_code)]
+
+fn gimme_static_mut() -> &'static mut u32 {
+    let ref mut x = 1234543; //~ ERROR borrowed value does not live long enough [E0597]
+    x
+}
+
+fn main() {}

--- a/src/test/ui/nll/return-ref-mut-issue-46557.stderr
+++ b/src/test/ui/nll/return-ref-mut-issue-46557.stderr
@@ -1,10 +1,10 @@
 error[E0597]: borrowed value does not live long enough
-  --> $DIR/return-ref-mut-issue-46557.rs:19:21
+  --> $DIR/return-ref-mut-issue-46557.rs:17:21
    |
-19 |     let ref mut x = 1234543; //~ ERROR borrowed value does not live long enough [E0597]
+17 |     let ref mut x = 1234543; //~ ERROR borrowed value does not live long enough [E0597]
    |                     ^^^^^^^ temporary value does not live long enough
-20 |     x
-21 | }
+18 |     x
+19 | }
    |  - temporary value only lives until here
    |
    = note: borrowed value must be valid for lifetime '_#2r...

--- a/src/test/ui/nll/return-ref-mut-issue-46557.stderr
+++ b/src/test/ui/nll/return-ref-mut-issue-46557.stderr
@@ -1,0 +1,13 @@
+error[E0597]: borrowed value does not live long enough
+  --> $DIR/return-ref-mut-issue-46557.rs:19:21
+   |
+19 |     let ref mut x = 1234543; //~ ERROR borrowed value does not live long enough [E0597]
+   |                     ^^^^^^^ temporary value does not live long enough
+20 |     x
+21 | }
+   |  - temporary value only lives until here
+   |
+   = note: borrowed value must be valid for lifetime '_#2r...
+
+error: aborting due to previous error
+

